### ISSUE TITLE
Type parameters of router actions do not match

### ISSUE THREAD

**Issue by livthomas** [ngrx/platform#561]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
</code></pre>

## What is the current behavior?
If you create a reducer which takes `action: RouterAction<MyState>` as a parameter, the type parameters in `RouterNavigationAction` and `RouterCancelAction` will be mismatched. `RouterNavigationAction` takes type parameter `T` for `routerState` while `RouterCancelAction` takes type parameter `T` for `storeState` and `V` for `routerState`. So if you define the type parameter in `RouterAction`, it will not derive `RouterNavigationAction` and `RouterCancelAction` correctly.

## Expected behavior:
`RouterNavigationPayload` and `RouterCancelPayload` should use the same type parameter for their `routerState` field.

## Minimal reproduction of the problem with instructions:
```typescript
export function navigationReducer(state: MyState, action: RouterAction<MyState>) {
  switch (action.type) {
    case ROUTER_NAVIGATION:
      // complains that it cannot be converted to RouterNavigationAction<RouterStateSnapshot>
      const navigationAction = action as RouterNavigationAction;
      const params = navigationAction.payload.routerState.root.paramMap;
      ...
    case ROUTER_CANCEL:
      ...
    default:
      return state;
  }
}
```

## Version of affected browser(s),operating system(s), npm, node and ngrx:
NgRx 4.1.1 with Angular 5.0.1 (and TypeScript 2.4.2)

**Comment by brandonroberts** [comment]
The type for the action should have 2 parameters, the first being the provided type for the snapshot and the second being the return type for the router actions.

```ts
export function navigationReducer(state: MyState, action: RouterAction<RouterStateSnapshot, MyState>) {
```

**Comment by livthomas** [comment]
@brandonroberts, isn't it vice versa? Because the second type parameter is optional with the default type `RouterStateSnapshot` (see [RouterAction definition](https://github.com/livthomas/ngrx-platform/blob/master/modules/router-store/src/router_store_module.ts#L88)).

**Comment by brandonroberts** [comment]
Both types are `RouterStateSnapshot`, the second is optional so it wasn't a breaking change to the existing type when a custom serializer was added.

**Comment by livthomas** [comment]
@brandonroberts, and what does the [storeState](https://github.com/livthomas/ngrx-platform/blob/master/modules/router-store/src/router_store_module.ts#L51) field represents then?

I can see the following explanation in the documentation:
> ROUTER_CANCEL and ROUTER_ERROR contain the store state before the navigation. Use the previous state to restore the consistency of the store.

**Comment by brandonroberts** [comment]
I misread your comment. Router-store uses an optimistic approach to updating the state when navigation occurs. T represents the shape of your store and V is the shape of the snapshot. Your PR does make that more clear.

**Comment by brandonroberts** [comment]
Closed via #562
